tests/ga/extension_actor.py

```diff
@@ -0,0 +1,465 @@
+        protocol._actors = [firstActor]
+        protocol._actors.extend(remainingActors)
+
+        protocol.set_http_handlers(http_get_handler=_generate_mock_http_get(protocol._actors[1:]),
+            http_put_handler=_generate_mock_http_put(protocol._actors))
+
+        try:
+            # enable the exthandlers_handler to interact with every actor before it is exposed.
+            for actor in protocol._actors:
+                patched_popen = actor.patch_popen() # We delay creation so that *this* patched_popen can "see" the prior ones (they must be started already).
+                patched_popen.start() # Allow later patches to "see" this one (as subprocess.Popen).
+
+                protocol._patched_popens.append(patched_popen)
+
+            yield protocol, get_exthandlers_handler(protocol)
+
+        finally:
+            for patched_popen in protocol._patched_popens:
+                patched_popen.stop()
+
+
+def add_extension_actors(protocol, incarnation, *actors):
+    """
+    Given a protocol obj returned by a extension_actor.get_protocol_and_handler(firstActor, *remainingActors) call, injects the
+    extensions emulated by (the objects within) actors.
+    
+    It accomplishes this though applying the following updates onto the provided protocol:
+        *   adds unique data sources for *.zip extension resources garnered from actors (unique meaning ones not already given
+            by [firstActor, *remainingActors], if any)
+        *   updates ext_conf to reflect the extension names and versions within actors
+        *   updates the goal state's incarnation (to force the agent to re-read it)
+
+    In order to force a goal state update, the incarnation parameter needs to be greater than the value passed in the last
+    extension_actor.update_extension_actors() function call, or greater than 0 if this is the first of such calls.
+
+    Note that non-default manifest.xml files must use the same url format as the default in order to be compatible with this
+    function. Specifically, the *.zip resource urls for an extension must end in "{ExtensionName}__{ExtensionVersion}".
+
+    Note that this function does not update xml for ext_conf or manifest, meaning that added extensions must already be present
+    within those files as provided by the firstActor's data source. This could probably be implemented with the
+    WireProtocolDataBase.replace_xml_element_value function, but it would probably make sense to first implement autogeneration
+    of those files within ExtensionActor.__init__, as we would then know exactly the xml elements in those files. This would
+    simplify the logic of adding the new actors, but is a bigger change. Ultimately, this feature isn't currently (i.e. as of
+    writing) needed.
+    """
+
+    new_actors = list(filter(lambda actor: actor not in protocol._actors, actors))  # list() applies the lambda before we edit protocol._actors
+    protocol._actors.extend(new_actors)
```

Comment:
attaching the actor to the protocol seemms odd, even if this test code. can we find a better place for them?

Reply:
I'm actually running into this with the "guaranteed ordering of command invocations" we mentioned earlier. I'm leaning towards some sort of "manager" class that would basically take on the three base-namespace functions (`get_extension_actor`, `get_protocol_and_handler`, and `add_extension_actors`) into itself as methods. Thoughts?